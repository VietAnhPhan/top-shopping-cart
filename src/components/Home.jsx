import { Link, Outlet } from "react-router-dom";
import Header from "./Header/Header";
import { useState } from "react";

function Home() {
  const [cart, setCart] = useState([]);

  return (
    <>
      <Header totalProductsCart={cart.length}></Header>
      <main>
        <h1>Welcome to our online store! We're so glad you're here.</h1>
        <p>
          Step into a world of endless possibilities, where quality meets
          convenience. Whether you're searching for the latest trends, everyday
          essentials, or unique gifts, we've carefully curated a collection that
          we know you'll love.
        </p>
        <Link to="/shop">Shop now</Link>
      </main>
      <Outlet context={{ cart, setCart }}></Outlet>
    </>
  );
}

export default Home;
