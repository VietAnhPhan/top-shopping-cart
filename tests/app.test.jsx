import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import Home from "../src/components/Home";
import {
  BrowserRouter,
  RouterProvider,
  createBrowserRouter,
} from "react-router-dom";
import routes from "../src/route";
import userEvent from "@testing-library/user-event";

describe("App component", () => {
  it("renders correct heading", () => {
    render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    );
    expect(screen.getByRole("heading").textContent).toMatch(
      /Welcome to our online store! We're so glad you're here./i
    );
  });

  it("renders product components", async () => {
    const router = createBrowserRouter(routes);
    render(<RouterProvider router={router}></RouterProvider>);

    const user = userEvent.setup();
    await user.click(screen.getAllByText(/Shop/i)[0]);
    await screen.findByRole("heading");
    const products = await screen.findAllByTestId("product-data-testid");
    await screen.findAllByTestId("product-title-data-testid");

    expect(screen.getByRole("heading")).toHaveTextContent("Our Products");
    expect(products.length).toBe(20);
    // expect(
    //   screen.getAllByTestId("product-title-data-testid")[0].textContent
    // ).toBe("Fjallraven - Foldsac....");
  });

  it("renders correct cart numbers", async () => {
    const router = createBrowserRouter(routes);
    render(<RouterProvider router={router}></RouterProvider>);
    const user = userEvent.setup();

    await user.click(screen.getAllByText(/Shop/i)[0]);

    await screen.findAllByRole("button", { name: "Add to Cart" });

    await user.click(screen.getAllByRole("button", { name: "Add to Cart" })[0]);
    await user.click(screen.getAllByRole("button", { name: "Add to Cart" })[1]);
    await user.click(screen.getAllByRole("button", { name: "Add to Cart" })[2]);

    expect(screen.getByTestId("cart-data-testid")).toHaveTextContent(
      "Cart (3)"
    );
  });

  it("renders correct total price", async () => {
    const router = createBrowserRouter(routes);
    render(<RouterProvider router={router}></RouterProvider>);
    const user = userEvent.setup();

    await user.click(screen.getAllByText(/Shop/i)[0]);
    await screen.findAllByRole("button", { name: "Add to Cart" });
    await user.click(screen.getAllByRole("button", { name: "Add to Cart" })[0]);
    await user.click(screen.getAllByRole("button", { name: "Add to Cart" })[1]);
    await user.click(screen.getAllByRole("button", { name: "Add to Cart" })[2]);

    const productPrices = screen.getAllByTestId("price");

    expect(
      parseFloat(productPrices[0].textContent.slice(1) * 2) +
        parseFloat(productPrices[1].textContent.slice(1) * 3) +
        parseFloat(productPrices[2].textContent.slice(1) * 4)
    ).toBe(510.76);
  });
});
