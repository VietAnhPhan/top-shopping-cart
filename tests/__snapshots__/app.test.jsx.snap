// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > navigation 1`] = `
<div>
  <header>
    <nav>
      <ul
        class="_menuItems_a5b977"
      >
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/cart"
          >
            Cart (
            0
            )
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main>
    <main>
      <h1>
        Welcome to our online store! We're so glad you're here.
      </h1>
      <p>
        Step into a world of endless possibilities, where quality meets convenience. Whether you're searching for the latest trends, everyday essentials, or unique gifts, we've carefully curated a collection that we know you'll love.
      </p>
      <a
        data-discover="true"
        href="/shop"
      >
        Shop now
      </a>
    </main>
  </main>
  <footer>
    <p>
      @2025 Anh Phan
    </p>
  </footer>
</div>
`;

exports[`App component > render 1`] = `
<div>
  <header>
    <nav>
      <ul
        class="_menuItems_a5b977"
      >
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/cart"
          >
            Cart (
            0
            )
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main>
    <main>
      <h1>
        Welcome to our online store! We're so glad you're here.
      </h1>
      <p>
        Step into a world of endless possibilities, where quality meets convenience. Whether you're searching for the latest trends, everyday essentials, or unique gifts, we've carefully curated a collection that we know you'll love.
      </p>
      <a
        data-discover="true"
        href="/shop"
      >
        Shop now
      </a>
    </main>
  </main>
  <footer>
    <p>
      @2025 Anh Phan
    </p>
  </footer>
</div>
`;

exports[`App component > render product 1`] = `
<div>
  <header>
    <nav>
      <ul
        class="_menuItems_a5b977"
      >
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/cart"
          >
            Cart (
            0
            )
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main>
    <main>
      <h1>
        Welcome to our online store! We're so glad you're here.
      </h1>
      <p>
        Step into a world of endless possibilities, where quality meets convenience. Whether you're searching for the latest trends, everyday essentials, or unique gifts, we've carefully curated a collection that we know you'll love.
      </p>
      <a
        data-discover="true"
        href="/shop"
      >
        Shop now
      </a>
    </main>
  </main>
  <footer>
    <p>
      @2025 Anh Phan
    </p>
  </footer>
</div>
`;

exports[`App component > render product components 1`] = `
<div>
  <header>
    <nav>
      <ul
        class="_menuItems_a5b977"
      >
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/cart"
          >
            Cart (
            0
            )
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main>
    <main>
      <h1>
        Welcome to our online store! We're so glad you're here.
      </h1>
      <p>
        Step into a world of endless possibilities, where quality meets convenience. Whether you're searching for the latest trends, everyday essentials, or unique gifts, we've carefully curated a collection that we know you'll love.
      </p>
      <a
        data-discover="true"
        href="/shop"
      >
        Shop now
      </a>
    </main>
  </main>
  <footer>
    <p>
      @2025 Anh Phan
    </p>
  </footer>
</div>
`;

exports[`App component > renders product components 1`] = `
<div>
  <header>
    <nav>
      <ul
        class="_menuItems_a5b977"
      >
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li
          class="_menuItem_a5b977"
        >
          <a
            data-discover="true"
            href="/cart"
          >
            Cart (
            0
            )
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main>
    <main>
      <h1>
        Welcome to our online store! We're so glad you're here.
      </h1>
      <p>
        Step into a world of endless possibilities, where quality meets convenience. Whether you're searching for the latest trends, everyday essentials, or unique gifts, we've carefully curated a collection that we know you'll love.
      </p>
      <a
        data-discover="true"
        href="/shop"
      >
        Shop now
      </a>
    </main>
  </main>
  <footer>
    <p>
      @2025 Anh Phan
    </p>
  </footer>
</div>
`;
